src/platform_impl/macos/view.rs

```diff
@@ -632,44 +632,91 @@ extern "C" fn key_down(this: &Object, _sel: Sel, event: id) {
+                } else {
+                    new_char = None;
+                }
+                AppState::queue_event(EventWrapper::StaticEvent(window_event));
+                match new_char {
+                    Some(_) => {
+                        AppState::queue_event(EventWrapper::StaticEvent(Event::WindowEvent {
+                            window_id,
+                            event: WindowEvent::ReceivedCharacter(new_char.unwrap()),
+                        }));
+                    }
+                    None => {
+                        for character in characters.chars().filter(|c| !is_corporate_character(*c))
+                        {
+                            AppState::queue_event(EventWrapper::StaticEvent(Event::WindowEvent {
+                                window_id,
+                                event: WindowEvent::ReceivedCharacter(character),
+                            }));
+                        }
+                    }
+                }
+            }
+            _ => {
+                #[allow(deprecated)]
+                let window_event = Event::WindowEvent {
+                    window_id,
+                    event: WindowEvent::KeyboardInput {
+                        device_id: DEVICE_ID,
+                        input: KeyboardInput {
+                            state: ElementState::Pressed,
+                            scancode,
+                            virtual_keycode,
+                            modifiers: event_mods(event),
+                        },
+                        is_synthetic: false,
+                    },
+                };
+
+                let pass_along = {
+                    AppState::queue_event(EventWrapper::StaticEvent(window_event));
+                    // Emit `ReceivedCharacter` for key repeats
+                    if is_repeat && state.is_key_down {
+                        for character in characters.chars().filter(|c| !is_corporate_character(*c))
+                        {
+                            AppState::queue_event(EventWrapper::StaticEvent(Event::WindowEvent {
+                                window_id,
+                                event: WindowEvent::ReceivedCharacter(character),
+                            }));
+                        }
+                        false
+                    } else {
+                        true
+                    }
+                };
+
+                if pass_along {
+                    // Some keys (and only *some*, with no known reason) don't trigger `insertText`, while others do...
+                    // So, we don't give repeats the opportunity to trigger that, since otherwise our hack will cause some
+                    // keys to generate twice as many characters.
+                    let array: id = msg_send![class!(NSArray), arrayWithObject: event];
+                    let _: () = msg_send![this, interpretKeyEvents: array];
```

Comment:
It seems to me that you're manually trying to regenerate things. This isn't the right way to solve this issue. IIRC it should be done with masks like iterm is doing here https://github.com/gnachman/iTerm2/blob/30e34ee31060e61787dc6e1968a487ab05d0aff5/sources/iTermKeyboardHandler.h#L16 . I guess tracking its code is a right way to go.

Reply:
Excellent! I'll research more about that approach. Thanks for the pointers.

Reply:
I don't have macOS systems tbh. But It's good to take a look what other terminals are doing here, since they should also solve this problem somehow.

Reply:
That's fine if you aren't using macOS, I'll try and represent a dev env. Note, if I follow the `winit` code correctly, it is translating the scan code to the virtual key-code since the retrieved character isn't in `char_to_keycode`. The scan code is supposed to independent of placement, yes? So, the big issue is with shift modifier? I'll still look into masks, I'm just making sure I follow the dependency on keyboard layouts.

Reply:
scancode is bound to place IIRC, this mapping is from Event.h(apple thing) though. But I could be wrong here, however you can verify that by using `dvorak` layout, I guess. So translating ascii chars to virtual key codes is the right way to go. However scancode for `return`, `backspace` and friends is likely identical regardless of the layout. Yes, shift modifier could be a problem here and for `Option` mode you're getting strange thing in the end anyway.

The problem is that we're getting strange characters from `Alt + a` input for e.g. so we should find out how to filter out them. I feel like Masks should do something about how macOS sending events to us. So the only way to go is to understand what this code in iterm is about and try to and clearly explain/port here.

Reply:
Okay, `Dvorak` works fine. However, `British - PC` has a different enough mapping to exhibit that once the `Shift` modifier is pressed, the incorrect key is sent to upstream. I will see if I can fix it.

Reply:
Okay, found it in Apple's documentation:

https://developer.apple.com/documentation/appkit/nsevent/1524605-charactersignoringmodifiers

Which makes much more sense.. Now, looking back at the code, I see that this can be accomplished very easily like so:

```patch
diff --git a/src/platform_impl/macos/view.rs b/src/platform_impl/macos/view.rs
index 741ecb4..7889898 100644
--- a/src/platform_impl/macos/view.rs
+++ b/src/platform_impl/macos/view.rs
@@ -510,10 +510,12 @@ extern "C" fn insert_text(this: &Object, _sel: Sel, string: id, _replacement_ran
         state.is_key_down = true;

         // We don't need this now, but it's here if that changes.
-        //let event: id = msg_send![NSApp(), currentEvent];
+        let event: id = msg_send![NSApp(), currentEvent];
+        let ev_mods = event_mods(event);
+        let raw_string = get_characters(event, ev_mods.alt());

         let mut events = VecDeque::with_capacity(characters.len());
-        for character in string.chars().filter(|c| !is_corporate_character(*c)) {
+        for character in raw_string.chars().filter(|c| !is_corporate_character(*c)) {
             events.push_back(EventWrapper::StaticEvent(Event::WindowEvent {
                 window_id: WindowId(get_window_id(state.ns_window)),
                 event: WindowEvent::ReceivedCharacter(character),
@@ -624,7 +626,8 @@ extern "C" fn key_down(this: &Object, _sel: Sel, event: id) {
         let state_ptr: *mut c_void = *this.get_ivar("winitState");
         let state = &mut *(state_ptr as *mut ViewState);
         let window_id = WindowId(get_window_id(state.ns_window));
-        let characters = get_characters(event, false);
+        let ev_mods = event_mods(event);
+        let characters = get_characters(event, ev_mods.alt());

         state.raw_characters = Some(characters.clone());
```

Of course, it still has to be cleaned up, because I don't understand the insertText event fully. It seems there is an interaction between keyDown and insertText, and it is a bit odd. Plus, we don't want to overwrite the string being passed in to insertText, right?

So many questions.. However, the above patch does produce the correct keys, and works as well with the `British - PC` layout. Plus, it appears key repeats work just fine, this allowed me to turn off most of my alacritty config, which is a big plus.

Reply:
>Plus, we don't want to overwrite the string being passed in to insertText, right?

I have no idea about this one, so someone who knows macOS should answer, I guess. But generally when you're not sure about something like that, I'd recommend to test and see what happens.

>Okay, found it in Apple's documentation:

Well, it says the thing we actually want in alacritty AFAICS. However we should keep old behavior under option, I guess, but let just try to implement something that works for us, and hide new behavior under option later during impl.

>Plus, it appears key repeats work just fine, this allowed me to turn off most of my alacritty config, which is a big plus.

Could you verify that some keys don't create twice as much events as they should on repeat? I'd recommend to press just all of them. I just wonder why you've mentioned key repeat here.

>So many questions.. However, the above patch does produce the correct keys, and works as well with the British - PC layout.

I'll assume that it just produces correct keys without mods or with shift, and no strange unicode thing when Alt is involved, right?. Have you verified Alt + Shift behavior.

Could you provide me `--print-events` log from alacritty during you pressing `s, shift + s, shift + alt + s, alt + s`? Just want to check.

Reply:
> > Plus, we don't want to overwrite the string being passed in to insertText, right?
> 
> I have no idea about this one, so someone who knows macOS should answer, I guess. But generally when you're not sure about something like that, I'd recommend to test and see what happens.

I have tested it, it works well. I support I mean more particularly, why use insertText at all? It causes the control flow to bounce around due to some key presses not triggering insertText. Copy and paste still work without insertText. (I tested this statement, I enjoy removing that pass_along style code.)

> > Okay, found it in Apple's documentation:
> 
> Well, it says the thing we actually want in alacritty AFAICS. However we should keep old behavior under option, I guess, but let just try to implement something that works for us, and hide new behavior under option later during impl.

Yes, I agree. I think we should close this pull request and open a new one with options to get this behavior. I'll go read how to get options into winit.

> > Plus, it appears key repeats work just fine, this allowed me to turn off most of my alacritty config, which is a big plus.
> 
> Could you verify that some keys don't create twice as much events as they should on repeat? I'd recommend to press just all of them. I just wonder why you've mentioned key repeat here.

At one point, I had key repeats not working. However, this is no longer reproducible. It may have been my hacks at the time while learning the project.

> > So many questions.. However, the above patch does produce the correct keys, and works as well with the British - PC layout.
> 
> I'll assume that it just produces correct keys without mods or with shift, and no strange unicode thing when Alt is involved, right?. Have you verified Alt + Shift behavior.
> 
> Could you provide me `--print-events` log from alacritty during you pressing `s, shift + s, shift + alt + s, alt + s`? Just want to check.

Absolutely. Trying to attach is failing.. let me see..

Reply:
Couldn't upload the file for some reason. You can read it here though:

https://gist.github.com/kjmph/f9b0b84e6ad9d62e1a329364a004f059

Reply:
Wow. `Alt-i` is very elucidating. It doesn't trigger `insert_text` since it isn't a printable character. Yet, when the event triggers `key_down` the `state` object doesn't have the `is_key_down` set to true, so it calls `interpretKeyEvents` and places the event back in the queue. Apple then interprets that as the circumflex diacritic, and will use it to append to the next character inserted. I went through the `winit` bugs, and there were a lot that referenced this code. By removing the `insert_text` and `do_command_by_selector`, a lot of code is cleaned up. I'm unable to reproduce any problems with this. I'll open a pull request showing these changes.

Reply:
I can't seem to invite you to review @kchibisov. Here is the improved pull request: https://github.com/rust-windowing/winit/pull/1449